### DISCUSSION TITLE
fail to hook function of 'system/lib/libmedia.so'

### DISCUSSION THREAD

**Discussion by JarYe** [bytedance/android-inline-hook#6] [Q&A] [answered]
### Error message :

E/shadowhook_tag: shadowhook: hook_sym_name(libmedia.so, _ZN7android10AudioTrack5writeEPKvjb, 0x71a36d82f0) FAILED. 18 - Find symbol in ELF failed

```cpp
 size_t my_write(const void* buffer, size_t userSize, bool blocking)
 {
     LOG("my_write proxy execute ...");
 }

void hookTest()
{
    stub = shadowhook_hook_sym_name("libmedia.so",
                                    "_ZN7android10AudioTrack5writeEPKvjb",
                                    (void*) my_write,
                                    (void**) &orig);
    if(stub == nullptr)
    {
        int err_num = shadowhook_get_errno();
        const char* err_msg = shadowhook_to_errmsg(err_num);
        LOG("hook sym name FAILED: %s", err_msg);
    }
}
```

**Comment by caikelun** [comment, marked as answer]
说明在 libmedia.so 中没有这个符号，可以用readelf确认下。

**Reply by JarYe** [reply]
![image](https://user-images.githubusercontent.com/20296408/160967038-c3bdd7f5-3870-4519-a099-0f50947750cf.png)

**Reply by caikelun** [reply]
UND表示这个符号不在当前so中，而是当前so需要调用的其他so中的符号。
建议可以先熟悉了解一下readelf的用法和各信息项的含义。

**Reply by JarYe** [reply]
感谢，已经注意到这个 UND

**Reply by JarYe** [reply]
找到真实地址了，还是报同样的错误。
E/shadowhook_tag: shadowhook: hook_sym_name(libaudioclient.so, _ZN7android10AudioTrack5writeEPKvjb, 0x71a36932f0) FAILED. 18 - Find symbol in ELF failed

![image](https://user-images.githubusercontent.com/20296408/161262946-5f4f8fde-7c2a-460a-802e-0d6fe37c67e1.png)

**Reply by caikelun** [reply]
请提供shadowhook版本号，Android OS版本号，准确的机型。

**Reply by JarYe** [reply]
shadownhook 版本： 1.0.3
vivo NEX S Android 10(Origin OS 1.0)

**Comment by caikelun** [comment]
你要注意，这个函数在Android 10源码中的定义是：
```
ssize_t AudioTrack::write(const void* buffer, size_t userSize, bool blocking)
```
其中的 size_t 在 32 位和 64 位上分别对应 int 和 long。所以这个函数在 arm32 / arm64 上的符号名不同：
arm32: _ZN7android10AudioTrack5writeEPKvjb
arm32: _ZN7android10AudioTrack5writeEPKvmb

**Reply by JarYe** [reply]
存在一个不就应该能 hook 中吗？如果两个不一样的情况下，需要怎么处理呢？

**Reply by caikelun** [reply]
```C
#if defined(__arm__)
    char *sym_name = "_ZN7android10AudioTrack5writeEPKvjb";
#elif defined(__aarch64__)
    char *sym_name = "_ZN7android10AudioTrack5writeEPKvmb";
#endif
```

**Reply by JarYe** [reply]
inline hook 应该理论上没有限制才对？
现在发现函数 hook 返回成功，但是最终没有执行。
往上层代码尝试，frameworks/av/media/libmedia/mediaplayer.cpp 这个路径下的可以 hook，但是回调会直接崩溃。
